hips/hip-9999.md

```diff
@@ -0,0 +1,72 @@
+---
+hip: 9999
+title: "Support Helm release data stored in multiple K8s Secrets"
+authors: [ "Ralf Van Dorsselaer" ]
+created: "2022-06-20"
+type: "feature"
+status: "draft"
+---
+ 
+## Abstract
+ 
+Helm release data encoded into the release field in a Kubernetes Secret can cause the Secret as a whole to exceed 1 MiB (as defined by MaxSecretSize in kubernetes/pkg/apis/core/types.go).
+ 
+When this happens, the Kubernetes Secret cannot be stored in Etcd and the Helm install or upgrade fails.
+ 
+Splitting the Helm release data across multiple Kubernetes Secrets resolves this problem.
+ 
+## Motivation
+ 
+Helm release data that causes the Kubernetes Secret size to exceed MaxSecretSize will make Helm install or upgrade fail.
+ 
+The existing HELM_DRIVER=secret storage driver does not handle Kubernetes Secrets become too big when storing a large encoded Helm release data.
+ 
+## Rationale
+ 
+The logical place to support this is in the HELM_DRIVER=secret storage driver.
+ 
+## Specification
+ 
+The feature is transparent to users. 
+ 
+When the HELM_DRIVER=secret storage driver notices that the resulting Kubernetes Secret will exceed MaxSecretSize then the storage driver will split the encoded release data across multiple Kubernetes Secrets.
+ 
+Splitting requires additional metadata to be added to the Kubernetes Secret:
+ 
+- number of chunks: integer, the number of Kubernetes Secrets that holds the release data for this Helm release version.
+- chunk: integer, the chunk number. Each Secret gets a chunk number. Reconstituting the release data can be done by reading the Secret with chunk=1, then subsequently reading in numerical order any number of additional Secrets up to and including the "number of chunks".
+ 
+Splitting requires the name of the Secret to include a "chunk" suffix. Example: sh.helm.release.v1.test.v1.2, 3, 4 and so on where the number after the last dot is the "number of chunk". The first Secret will not get a .1 suffix.
+ 
+If no splitting is required then no additional metadata is included/set.
+ 
+## Backwards compatibility
+ 
+If the chunking is implemented in the secret/secrets HELM_DRIVER then backward compatibility to previously installed Helm releases is guaranteed. If no chunking information is encoded in the release secret then the driver can behave as before chunking support was added. If no chunking is needed then the release secret can be created without any additional metadata and with compatible name.
```

Comment:
What about compatibility with third-party tooling? For example https://github.com/helm/helm-mapkubeapis.

Reply:
What I understand from looking at the https://github.com/helm/helm-mapkubeapis code is that it is not impacted as it works as a Helm plugin and its code seems to be agnostic to how the Helm release is actually stored.

For tools/scripts that read K8s Secrets directly (and are expecting to find a single encoded blob that describes the Helm release) things are different.

In the scenario where the encoded release data size is < (1000 *1024) bytes there is no problem for third parties to read and parse the Helm release data. Only if the Helm release data is split into multiple parts these tools/scripts would be broken.

I have following question:

- Is the Helm release information encoded into a K8s Secret considered a public interface? Is there public documentation from the Helm project that describes the format of the data inside a Kubernetes Secret? Is the Helm project encouraging reading K8s Secrets directly and developing tooling around this?

If the internal format is not documented publicly and considered a private interface then any third party tool that relies on it will have reverse engineered it from looking at K8s Secrets and Helm code and is therefore vulnerable to any changes to the internal format.

If the internal format is publicly documented and considered a public interface then that does not change much because any change to support storing a Helm release differently in K8s Secrets would result in incompatible changes for consuming third parties. The breaking change for the Helm release storage format for said third parties would have to be communicated early as part of the major release change log.

Reply:
While the internal format is not documented publicly, it is expected that older versions of Helm 3's SDK shall continue to work up until the release of Helm 4. If the format of the secret changes to the point that older versions of Helm 3 do not support read/write operations against it, that is considered a breaking change.

If we allow the storage of releases larger than 1MB, older versions of the Helm 3 SDK is incompatible with the release secrets.

This is why we've generally steered users towards [the database driver](https://helm.sh/docs/topics/advanced/#sql-storage-backend) for Helm 3, because it is documented to support release types larger than 1MB in size.

Reply:
Is that a realistic expectation that any old 3.x client works with any release of any chart?  We have been publishing various releases of our product as a helm chart over the past years, and I recall having to advice customers several times that we need a certain minimum version of helm, e.g. to avoid running into some defect that was fixed only in a newer version - and that did not create problems with customers, they seemed to be on reasonably up-to-date versions anyway. 

As far as I understand your concern, such a worst-case corruption problem might only occur for an application that depends on the chunking support (i.e. a respective version of helm), and there is no risk for a broader corruption at the overall cluster.  In other words, if a helm-based applications has a large chart, it needs to be the application's concern that they need to instruct their customers to be on a respective version of helm, and as said, we had to that anyway in the past because of defect fixes we depended on.